src/index/ivf/ivf.cc

```diff
@@ -138,30 +134,18 @@ class IvfIndexNode : public IndexNode {
     Type() const override {
         if constexpr (std::is_same<T, faiss::IndexIVFFlat>::value) {
             return knowhere::IndexEnum::INDEX_FAISS_IVFFLAT;
-        }
-        if constexpr (std::is_same<T, faiss::IndexIVFPQ>::value) {
+        } else if constexpr (std::is_same<T, faiss::IndexIVFPQ>::value) {
             return knowhere::IndexEnum::INDEX_FAISS_IVFPQ;
-        }
-        if constexpr (std::is_same<T, faiss::IndexIVFScalarQuantizer>::value) {
+        } else if constexpr (std::is_same<T, faiss::IndexIVFScalarQuantizer>::value) {
             return knowhere::IndexEnum::INDEX_FAISS_IVFSQ8;
-        }
-        if constexpr (std::is_same<T, faiss::IndexBinaryIVF>::value) {
+        } else if constexpr (std::is_same<T, faiss::IndexBinaryIVF>::value) {
             return knowhere::IndexEnum::INDEX_FAISS_BIN_IVFFLAT;
         }
     };
-    ~IvfIndexNode() override {
-        if (index_) {
-            delete index_;
-        }
-        if (qzr_) {
-            delete qzr_;
-        }
-    };
 
  private:
-    T* index_;
+    std::shared_ptr<T> index_;
```

Comment:
use T * ptr cause memory leak?

Reply:
There are many places we need check "index_ != nullptr" then do "delete index_", including exception handling and gpu index / cpu index copying, 
if use shared_ptr, we can make coding easier and leak free.